langkit/utils.py

```diff
@@ -1,11 +1,38 @@
 import os
-from typing import Optional
+from typing import Optional, Generic, Callable, Dict, TypeVar
 from langkit import lang_config
 import string
 import random
 import functools
 import warnings
 
+Out = TypeVar("Out")
+In = TypeVar("In")
+
+
+class DynamicLazyInit(Generic[In, Out]):
+    def __init__(self, init: Callable[[In], Out]):
+        self.__cache: Dict[In, Out] = {}
+        self.__init = init
+
+    def value(self, arg: In) -> Out:
+        if arg not in self.__cache:
+            self.__cache[arg] = self.__init(arg)
+
+        return self.__cache[arg]
+
+
+class LazyInit(Generic[Out]):
```

Comment:
Can we prefix with _ to signify these are internal.

Reply:
removed those, not relevant anymore